apollo-client/src/main/java/com/ctrip/framework/apollo/spring/config/ConfigPropertySource.java

```diff
@@ -16,32 +16,29 @@
  */
 package com.ctrip.framework.apollo.spring.config;
 
+import com.ctrip.framework.apollo.Config;
 import com.ctrip.framework.apollo.ConfigChangeListener;
-import java.util.Set;
-
 import org.springframework.core.env.EnumerablePropertySource;
 
-import com.ctrip.framework.apollo.Config;
-
 /**
  * Property source wrapper for Config
  *
  * @author Jason Song(song_s@ctrip.com)
  */
 public class ConfigPropertySource extends EnumerablePropertySource<Config> {
-  private static final String[] EMPTY_ARRAY = new String[0];
 
   ConfigPropertySource(String name, Config source) {
     super(name, source);
   }
 
+  @Override
+  public boolean containsProperty(String name) {
+    return this.source.getProperty(name, null) != null;
+  }
+
   @Override
   public String[] getPropertyNames() {
-    Set<String> propertyNames = this.source.getPropertyNames();
-    if (propertyNames.isEmpty()) {
-      return EMPTY_ARRAY;
-    }
-    return propertyNames.toArray(new String[propertyNames.size()]);
+    return this.source.getPropertyNames();
```

Comment:
How about we keep everything else unchanged and only cache the property names as array in ConfigPropertySource? Then we could use ConfigChangeListener to update the cache.

Reply:
That's a good idea, even better!

Reply:
> only cache the property names as array in ConfigPropertySource

Or maybe we could move one step further, only cache the property names as array in `CompositeConfigPropertySource`.

Reply:
I think it's still better to cache the property names as array in ConfigPropertySource since it's also used in other places like `com.ctrip.framework.apollo.config.data.importer.ApolloConfigDataLoader`.